enhancements/cnf-tests-split.md

```diff
@@ -0,0 +1,204 @@
+---
+title: cnf-tests-split
+authors:
+  - "@liornoy"
+reviewers:
+  - "@cnf-network-team"
+  - "@cnf-ci-team"
+  - "@cnf-compute-team"
+approvers:
+  - "@cnf-network-team"
+  - "@cnf-ci-team"
+  - "@cnf-compute-team"
+creation-date: 2023-06-01
+last-updated: 2023-06-01
+tracking-link:
+  - https://issues.redhat.com/browse/CNF-7583
+---
+
+# CNF-tests split
+
+The cnf-tests is a sub-project in the cnf-features-deploy repository, aimed to run tests against various CNF workloads on top of an Openshift cluster.  
+The cnf-tests are running on upstream and downstream CI in a periodic and nightly manner in order to verify that the workloads are functioning as intended.
+
+## Summary
+
+The purpose of this enhancement is to decouple the cnf-tests from the test suites of the external CNF projects. Additionally, as this change affects the way tests are run, we propose rewriting the cnf-tests container image, which is a supported product for running the latency tests.
+## Motivation
+
+In the cnf-tests, all three test suites (configsuite, e2esuite, valdiationsuite) are importing
+external test suites (outside of the cnf-feauters-deploy repository) with a [blank identifier](https://golangbyexample.com/blank-identifier-import-golang/). This is done to compile their tests into one binary. For example, in the e2esuite, we import the following:
+
+```go
+import (
+    _ "github.com/k8snetworkplumbingwg/sriov-network-operator/test/conformance/tests"
+    _ "github.com/metallb/metallb-operator/test/e2e/functional/tests"
+    _ "github.com/openshift/cluster-node-tuning-operator/test/e2e/performanceprofile/functests/1_performance" 
+    _ "github.com/openshift/cluster-node-tuning-operator/test/e2e/performanceprofile/functests/4_latency" 
+    _ "github.com/openshift/ptp-operator/test/conformance/serial"
+)
+```
+
+To prevent conflicts due to the numerous dependencies in the test suites, we want to decouple the external test suites by introducing a script that invokes the required test suites as git-submodules under the cnf-tests/submodules path.
+### Note
+1. In this proposal, we refer to each external test suite as a CNF component.
+2. Some dependencies will still be required in the components as there are tests that combine multiple features.
+
+
+This change will also allow us to pin the external repositories to the latest branch, ensuring that we always test the latest tests without the need for manual commit hash updates. We will include an option to set an environment variable that specifies the commit ref for each repo.
```

Comment:
I remember from the meetings we had that we did want to always test the latest, and being pinned to a specific commit was a disadvantage.
You are right that it will cause failures to show up more frequently (ci fails more frequently), but isn't it the point?
to catch it early and fix it before shipping it to the customers?
I would like to hear more opinions about that please: @cgoncalves @yanirq @sshnaidm

Reply:
I don't remember this discussion, and not having control of what is running where is a recipe for disaster. Executions must be consistent as much as we can.
It's not about the frequency of the failures, is about knowing what tests are running and being able to reproduce.

Reply:
I'm not sure I understand the point of manual pinning. What is the criteria for update then?
If it's passing all tests, then you anyway should run CI jobs on it.

Reply:
Well, somebody should pass CI ideally for committing the test, and that supposes to block breakages. But I see your point, it can be a flaky test for example. I'm not against having a pinning mechanism for returning to a passing point, but not sure it will be updating all the time. People usually submit test and go. We have a real danger to run old and probably not relevant tests because nobody updates it here.
But it's always good to have option to pin it to a previous version where tests were passing until it's fixed.

Reply:
Maybe by combining both approaches, we will both pin to a specific commit to be on the safe side, plus
create an automated bump script like we have in metallb-operator and thus we could catch if something is failing in the PR's ci.
This script will save us time updating all the time and will create the bump PR for us automatically.

Reply:
Let me try to expand. With the current version, we are in control of what test ran in what day (because we bump the tests manually). This allows us to have a reproducible environment when a given test fails, which allows us to understand if the issue was caused by a test or by the feature that we want to test.

On top of that, given we gate (at least some features) with CI, bumping the tests with a PR will allow us to catch those issues early, because the bump PR itself will fail, giving a way more narrowed scope compared to the periodic CI. This is better, imo, to having to triage the periodic CI that starts to fail.

This requires the same effort that we put in today, by aligning each feature, with the advantage that we can automate the mechanism as @liornoy said, because now it's about bumping a commit instead of having to deal with go dependencies.

I'd also like to hear @SchSeba 's view here

Reply:
I understand the need of stable CI in this repo or wherever. For Upstream CI it's not the case, we need to latest code to use for tests, usually this code contains latest fixes as well. It doesn't make sense to run CI on the stable code, its purpose to run on the most fresh in order to catch issues ASAP and as close to the code as possible. Running on old code is something else, not CI.
Although it might work with automated PRs, but I very doubt it, probably it will include both fixes for old issues and new bugs, so it might be blocked for updates for a long time, also someone will need to maintain it, which is might be a problem by itself in this repo that always needed more love.
Anyway, I don't mind to have any updating mechanisms here, but for our CI we must have the latest code, if it fails - then bugs should be submitted and resolved (and probably tests skipped until resolved). If it will be submodules, I probably can clone the repo and force submodules to fetch the most fresh code instead of pinned one. If it will be done in a different way, then we'll need other ways to have the newest code. But this is a **hard requirement** from our side.

Reply:
> I understand the need of stable CI in this repo or wherever. For Upstream CI it's not the case, we need to latest code to use for tests, usually this code contains latest fixes as well. It doesn't make sense to run CI on the stable code, its purpose to run on the most fresh in order to catch issues ASAP and as close to the code as possible. 

I agree. What I am advocating here is not because of stability, it is because of consistency and reproducibility
Running against "main" means "we don't know what version of tests we are running""

> Running on old code is something else, not CI. Although it might work with automated PRs, but I very doubt it, probably it will include both fixes for old issues and new bugs, so it might be blocked for updates for a long time, also someone will need to maintain it, which is might be a problem by itself in this repo that always needed more love

This is what I was saying above. If a PR that bumps the tests in a way that makes them fail, it's better to know from the PR and not from the failing CI, because it's clearer to know what is causing the failure. Not pinning means hiding this under the carpet and making it harder to find that a change in the tests broke them.

>  Anyway, I don't mind to have any updating mechanisms here, but for our CI we must have the latest code, if it fails - then bugs should be submitted and resolved (and probably tests skipped until resolved). If it will be submodules, I probably can clone the repo and force submodules to fetch the most fresh code instead of pinned one. If it will be done in a different way, then we'll need other ways to have the newest code. But this is a **hard requirement** from our side.


If downstream CI will override the pinning then there's no point in the discussion. It will just make triaging failures  harder for the reasons I stated above.

Reply:
> I agree. What I am advocating here is not because of stability, it is because of consistency and reproducibility Running against "main" means "we don't know what version of tests we are running""
> 

I'm not sure why it's not possible to know what version of tests we are running. If it's a technical problem, it can be solved easily saving commit hashes somewhere. Jenkins does it automatically, and in Prow we'll need probably a primitive script to do it. Later by looking between hashes of success and failure we can see where the problem is. Isn't it as usually do?
As I understood the main problem for you is that while investigating and trying to fix the bug, the CI will continue to fail. Correct me if I'm wrong.
In our case we're implementing skipping of failing tests.

> 
> If downstream CI will override the pinning then there's no point in the discussion. It will just make triaging failures harder for the reasons I stated above.

I don't know if they do, but I discovered recently downstream CI uses `nto` repository from upstream for example, and that provides them fixes faster than we have in "upstream CI", because we're stuck here with old versions... That's weird.
cc @yanirq

Reply:
I just spotted an issue, and I don't know if that's a deal breaker, and it's just some technical thing, but if we agreed to keep the cnf-tests docs (DOCGEN stuff) then it must be updated via PR every time a new test is added/removed/updated.
Pinning to the main branch will cause the `make validate-test-list` to fail without properly editing the `TESTLIST.md`. correct me if I'm wrong.

Reply:
> As I understood the main problem for you is that while investigating and trying to fix the bug, the CI will continue to fail. Correct me if I'm wrong.

Nope, if there's a failure and there's a bug related to that failure, I don't mind if that test keeps failing or not, because we are aware there's an issue to be triaged.

My main problem is not being in control of what runs (and what ran, at a given point in time), and about swallow whatever change is applied to the tests before validating them, and getting a "all tests failed" report just because the location of the e2e tests moved in some repo we depend on.

The "what ran" can be solved by dumping the commit hashes as you said, the second part probably not, but we'll live with that I guess. 
Note that this also change the semantics of this repo, which becomes just glue scripts for running single test suites in an uniform matter, rather than a stand alone entity that can be consumed at any given point in time, because every time we run it, it may run a different set of tests depending on what "latest" contains.

That goes together with @liornoy point of not being able to document the tests contained by the suite.

Said that, I don't want to hold the design on this item. As long as we are able to know what version of the tests ran and what version of each feature was deployed at a given time, I am good.